### PR TITLE
Add a Cython-level cancel-many-calls test

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
I wrote this in the course of looking for a bug and then wasn't able to find the bug but oh look here's this test we might as well have with us going forward.

### CODE REVIEW

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                _EMPTY_METADATA, _EMPTY_FLAGS),
+            cygrpc.operation_send_message(b'\x79\x57', _EMPTY_FLAGS),
+            cygrpc.operation_send_status_from_server(
+                _EMPTY_METADATA, cygrpc.StatusCode.ok, b'test details!',
+                _EMPTY_FLAGS),
+        )
+        self._call.start_batch(
+            cygrpc.Operations(operations), _SERVER_COMPLETE_CALL_TAG)
+      self._completion_queue.poll()
+      self._completion_queue.poll()
```
Comment (kpayson64) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: What does the second poll do?

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+
+def _serve(state, server, server_completion_queue, thread_pool):
+  for _ in range(test_constants.RPC_CONCURRENCY):
+    with state.condition:
```
Comment (kpayson64) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: Do you need to acquire the lock here?

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    state = _State()
+
+    server_thread_args = (
+        state, server, server_completion_queue, server_thread_pool,)
+    server_thread = threading.Thread(target=_serve, args=server_thread_args)
+    server_thread.start()
+
+    client_condition = threading.Condition()
+    client_due = set()
+    client_completion_queue = cygrpc.CompletionQueue()
```
Comment (kpayson64) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: Its my understanding that from the client side we want a 1:1 call<->completion queue.  Do we have a single completion queue in Python?

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            cygrpc.operation_receive_message(_EMPTY_FLAGS),
+            cygrpc.operation_receive_status_on_client(_EMPTY_FLAGS),
+        )
+        tag = 'client_complete_call_{0:04d}_tag'.format(index)
+        client_call.start_batch(cygrpc.Operations(operations), tag)
+        client_due.add(tag)
+        client_calls.append(client_call)
+
+    with state.condition:
+      while True:
```
Comment (kpayson64) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: Not sure if this is a valid concern, but if parked_handlers reaches the target value before handled_rpcs (or vice versa) this becomes a spin loop.

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: Pulls the last event due off the completion queue. I'm not sure if it's guaranteed which of `_RECEIVE_CLOSE_ON_SERVER_TAG`, `_RECEIVE_MESSAGE_TAG`, or `_SERVER_COMPLETE_CALL_TAG` will be the last to come out, but the invariant for the whole method is that all tags that go into the underlying system are pulled out.

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py (hunk 2 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: Probably not.

File: src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py (hunk 3 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_cython/_cancel_many_calls_test.py]: I think we want whatever number of completion queues allow us to implement correct and performant code. In the case of this test there is only one completion queue on the client-side and that's just fine.

**Comment by nathanielmanistaatgoogle** [comment]
Comments addressed; back to you.

**Comment by kpayson64** [comment]
LGTM

**Pull request merged**